x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/PreAnalyzer.java

```diff
@@ -43,17 +44,19 @@ public PreAnalysis preAnalyze(LogicalPlan plan) {
     }
 
     protected PreAnalysis doPreAnalyze(LogicalPlan plan) {
-        Holder<IndexMode> indexMode = new Holder<>();
-        Holder<IndexPattern> indexPattern = new Holder<>();
+        Map<IndexPattern, IndexMode> indexes = new HashMap<>();
         List<IndexPattern> lookupIndices = new ArrayList<>();
         plan.forEachUp(UnresolvedRelation.class, p -> {
             if (p.indexMode() == IndexMode.LOOKUP) {
                 lookupIndices.add(p.indexPattern());
-            } else if (indexMode.get() == null || indexMode.get() == p.indexMode()) {
-                indexMode.set(p.indexMode());
-                indexPattern.set(p.indexPattern());
+            } else if (indexes.containsKey(p.indexPattern()) == false || indexes.get(p.indexPattern()) == p.indexMode()) {
+                indexes.put(p.indexPattern(), p.indexMode());
             } else {
-                throw new IllegalStateException("index mode is already set");
+                IndexMode m1 = p.indexMode();
+                IndexMode m2 = indexes.get(p.indexPattern());
+                throw new IllegalStateException(
+                    "index pattern '" + p.indexPattern() + "' found with with different index mode: " + m2 + " != " + m1
+                );
```

Comment:
I believe IndexMode is read from the index settings. Is it even possible to find the same pattern in 2 different index modes?

Reply:
OK. Let's try use an assert.